File: programmers/js/level01/17682.js
// [1차] 다트 게임 
function solution(dartResult) {
  const result = [];
  
  for (const dart of dartResult.match(/(\d+)([SDT])([*#]*)/g)) {
    let [num, area, option] = dart.split(/([SDT])/);

    if (area === 'S') area = 1;
    if (area === 'D') area = 2;
    if (area === 'T') area = 3;
    if (option === '') option = 1;
    if (option === '#') option = -1;
    if (option === '*') option = result.length ? 
      (result[result.length - 1] *= 2, 2) : 2;
    
    result.push(num ** area * option);
  }
  
  return result.reduce((acc, cur) => acc + cur, 0);
}

File: programmers/java/level01/17682.java
// [1차] 다트 게임

class Solution {
    public int solution(String dartResult) {
        int result = 0;
        int prevScore = 0;
        int currentScore = 0;
        String temp = "";
        
        for (int i = 0; i < dartResult.length(); i++) {
            char value = dartResult.charAt(i);
            
            if (value == 'S' || value == 'D' || value == 'T') {
                prevScore = currentScore;
                currentScore = Integer.parseInt(temp);
                
                if (value == 'D') currentScore *= currentScore;
                else if (value == 'T') currentScore *= currentScore * currentScore;
                
                result += currentScore;
                temp = "";
            } else if (value == '#') {
                currentScore = -currentScore;
                result += currentScore * 2;
            } else if (value == '*') {
                result += currentScore + prevScore;
                currentScore = currentScore * 2;
            } else {
                temp += value;
            }
        }
        
        
        return result;
    }
}